tidy3d/web/api/container.py

```diff
@@ -176,25 +170,19 @@ def run(self, path: str = DEFAULT_DATA_PATH) -> SimulationDataType:
             Object containing simulation results.
         """
 
+        self.upload()
         self.start()
         self.monitor()
         return self.load(path=path)
 
-    @pd.root_validator()
-    def _upload(cls, values) -> None:
-        """Upload simulation to server without running."""
-
-        # task_id already present, don't re-upload
-        if values.get("task_id") is not None:
-            return values
+    @cached_property
+    def task_id(self):
+        upload_kwargs = {key: getattr(self, key) for key in self._upload_fields}
+        return web.upload(**upload_kwargs)
 
+    def upload(self):
         # upload kwargs with all fields except task_id
-        upload_kwargs = {key: values.get(key) for key in cls._upload_fields}
-        task_id = web.upload(**upload_kwargs)
-
-        # then set the task_id and return
-        values["task_id"] = task_id
-        return values
+        self.task_id
```

Comment:
the function with upload but just directly return sefl.task_id will confuse the reader very much.

Reply:
ok how about I change so that

```py
    @cached_property
    def task_id(self)  -> TaskID:
        return self.upload()

    def upload(self) -> TaskID:
        # upload kwargs with all fields except task_id
        upload_kwargs = {key: getattr(self, key) for key in self._upload_fields}
        return web.upload(**upload_kwargs)
```

the idea is that when `.task_id` is asked for, if it hasn't been computed, it will `upload()` the task.